### PR TITLE
Phase control: remote active power mode

### PR DESCRIPTION
**Pull request by tadam50**
Signed-off-by: Thomas ADAM <tadam@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#84 

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java [hunk 1]
```diff
@@ -46,25 +49,73 @@ public static LfLegBranch create(LfBus bus1, LfBus bus0, ThreeWindingsTransforme
...
         double nominalV2 = twt.getRatedU0();
         double zb = nominalV2 * nominalV2 / PerUnit.SB;
         PhaseTapChanger ptc = leg.getPhaseTapChanger();
         if (ptc != null
                 && ptc.isRegulating()
                 && ptc.getRegulationMode() != PhaseTapChanger.RegulationMode.FIXED_TAP) {
-            LOGGER.error("3 windings transformer '{}' has a regulating phase tap changer which is not yet supported", twt.getId());
+            PhaseTapChanger.RegulationMode regulationMode = ptc.getRegulationMode();
+            PhaseControl.ControlledSide controlledSide = null;
+            if (ptc.getRegulationTerminal() == twt.getLeg1().getTerminal()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java:65]: I think the code only allows to regulate locally. It means that a phase tap changer located in a dedicated leg can only regulated at its local terminal. I think that next week I will increase the regulation capability to the 2 other terminals of the transformer (but I don't know if it is easy of not ;-)).

File: src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java [hunk 2]
```diff
@@ -46,25 +49,73 @@ public static LfLegBranch create(LfBus bus1, LfBus bus0, ThreeWindingsTransforme
...
+                controlledSide = PhaseControl.ControlledSide.ONE;
+            } else if (ptc.getRegulationTerminal() == twt.getLeg2().getTerminal()) {
+                controlledSide = PhaseControl.ControlledSide.TWO;
+            } else if (ptc.getRegulationTerminal() == twt.getLeg3().getTerminal()) {
+                controlledSide = PhaseControl.ControlledSide.THREE;
+            } else {
+                LOGGER.error("3 windings transformer '{}' has a regulating phase tap changer which is not yet supported", twt.getId());
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java:72]: I think that here the best solution could be to consider the phase tap changer as ````FIXED_TAP````. If we do that choice, we have to modify ````LfBranchImpl```` too.

File: src/test/java/com/powsybl/openloadflow/ac/AcLoadFlowPhaseShifterTest.java [hunk 3]
```diff
@@ -62,7 +54,9 @@ void setUp() {
     }
 
     @Test
-    void baseCaseTest() {
+    void baseCase2WTTest() {
```
Comment (annetill) [inline src/test/java/com/powsybl/openloadflow/ac/AcLoadFlowPhaseShifterTest.java:57]: Can you change for ````baseCaseT2wtTest()```` ? I have forgotten.

**Review by annetill — COMMENTED** [review]
Just some comments, good beginning!

**Comment by floriand-e2r** [comment]
@tadam50 After discussing with @annetill the PhaseControl design, I refactored a bit your proposition:
- both controller and controlee now hold a reference to the corresponding phase control; indeed we often need only the control of the controlee, and going each time through the controlled branch seemed tedious to me 
- the phase control is now created at once (was previously created in the controller branch and after updated)
- even though phase control and voltage control share the controller/controlee principle, they don't have any common methods/attributes, hence I don't think inheritance from a common class helped

**Comment by annetill** [comment]
@floriand-e2r I think that the limiter mode is more complex than expected. So, what do you think about removing it from this PR and open a new PR for only that feature ? So, this PR will be mergeable ans it will be easier to work on discrete voltage control (#127 ).

File: src/main/java/com/powsybl/openloadflow/network/impl/LfBranchImpl.java [hunk 4]
```diff
@@ -185,34 +163,31 @@ public double getPermanentLimit2() {
...
-
     @Override
     public void updateState() {
         branch.getTerminal1().setP(p1.eval() * PerUnit.SB);
         branch.getTerminal1().setQ(q1.eval() * PerUnit.SB);
         branch.getTerminal2().setP(p2.eval() * PerUnit.SB);
         branch.getTerminal2().setQ(q2.eval() * PerUnit.SB);
 
-        if (phaseControl != null) { // it means there is a regulating phase tap changer
+        if (isPhaseController()) { // it means there is a regulating phase tap changer
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfBranchImpl.java:173]: @floriand-e2r Do not forget to add condition && phaseControl.getMode() == PhaseControl.Mode.OFF

File: src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java [hunk 5]
```diff
@@ -133,10 +155,20 @@ public void updateState() {
         leg.getTerminal().setP(p.eval() * PerUnit.SB);
         leg.getTerminal().setQ(q.eval() * PerUnit.SB);
 
-        if (phaseControl != null) {
+        if (isPhaseController()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfLegBranch.java:158]: @floriand-e2r Do not forget to add condition && phaseControl.getMode() == PhaseControl.Mode.OFF

File: src/main/java/com/powsybl/openloadflow/ac/PhaseControlOuterLoop.java [hunk 6]
```diff
@@ -33,47 +33,72 @@ public String getType() {
...
-                    a1.setActive(false);
-
-                    // de-activate phase control equation
-                    Equation t = context.getEquationSystem().createEquation(branch.getNum(), EquationType.BRANCH_P);
-                    t.setActive(false);
+        switch (context.getIteration()) {
+            case 0:
+                return checkAtFirstIteration(context);
+            case 1:
+                return checkAtSecondIteration(context);
```
Comment (floriand-e2r) [inline src/main/java/com/powsybl/openloadflow/ac/PhaseControlOuterLoop.java:41]: I'm not so happy with those names I gave... any suggestions?

File: src/main/java/com/powsybl/openloadflow/ac/PhaseControlOuterLoop.java [hunk 7]
```diff
@@ -33,47 +33,72 @@ public String getType() {
 
     @Override
     public OuterLoopStatus check(OuterLoopContext context) {
```
Comment (floriand-e2r) [inline src/main/java/com/powsybl/openloadflow/ac/PhaseControlOuterLoop.java:35]: In this outerloop we do more than just check in this method: we also switch off/on phase control. Shouldn't it be done in another method, or is it just a naming issue?

**Pull request merged**